src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -60,8 +60,17 @@ public static void Resize<T>([NotNull] ref T[]? array, int newSize)
 
             if (larray.Length != newSize)
             {
+                // Due to array variance, it's possible that the incoming array is
+                // actually of type U[], where U:T; or that an int[] <-> uint[] or
+                // similar cast has occurred. In any case, since it's always legal
+                // to reinterpret U as T in this scenario (but not necessarily the
+                // other way around), we can use Buffer.Memmove here.
+
                 T[] newArray = new T[newSize];
-                Copy(larray, 0, newArray, 0, larray.Length > newSize ? newSize : larray.Length);
+                Buffer.Memmove(
+                    ref MemoryMarshal.GetArrayDataReference(newArray),
+                    ref MemoryMarshal.GetArrayDataReference(larray),
+                    (uint)Math.Min(newSize, larray.Length));
```

Comment:
I'm not clear on why it's ok to use Buffer.Memmove rather than potentially needing to use Buffer.BulkMoveWithWriteBarrier?

Reply:
`Buffer.Memmove` will still call `BulkMoveWithWriteBarrier` when _T_ is a reference type or contains references. See:

https://github.com/dotnet/runtime/blob/a97e47483409bc8e02c302c44e9a395ab019b180/src/libraries/System.Private.CoreLib/src/System/Buffer.cs#L318-L338

What I was really trying to point out is that say we have the following code:

```cs
Base[] arr = new Derived[] { /* ... */ };
Array.Resize<Base>(ref arr, 20);
```

then this method will internally call `new Base[20]`. But since clearly instances of `Derived` can be placed into an array of type `Base[]`, it also follows that the entire contents of `Derived[]` can be quickly moved to an array of type `Base[]` without doing all of the method table checks that `Array.Copy` normally does. After all, that's what `new ReadOnlySpan<Base>(new Derived[] { /* ... */}).Slice(/* ... */).ToArray()` would've done. And `ReadOnlySpan<T>` (unlike `Span<T>`) doesn't bother with any array variance checks.

https://github.com/dotnet/runtime/blob/a97e47483409bc8e02c302c44e9a395ab019b180/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L45-L55

https://github.com/dotnet/runtime/blob/a97e47483409bc8e02c302c44e9a395ab019b180/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L384-L392

Reply:
Ah, I misread the code... I thought you were calling the byte overload.